src/main/java/org/radarbase/appserver/service/FcmNotificationService.java

```diff
@@ -1,22 +1,311 @@
+
+        if(notification.isEmpty()) {
+            throw new NotFoundException("Notification does not exist. Please create first");
+        }
+
+        Notification newNotification = notification.get()
+                .setBody(notificationDto.getBody())
+                .setScheduledTime(notificationDto.getScheduledTime().toInstant(ZoneOffset.UTC))
+                .setSourceId(notificationDto.getSourceId())
+                .setTitle(notificationDto.getTitle())
+                .setTtlSeconds(notificationDto.getTtlSeconds())
+                .setType(notificationDto.getType())
+                .setUser(user.get())
+                .setFcmMessageId(String.valueOf(notificationDto.hashCode()));
+        newNotification = this.notificationRepository.save(newNotification);
+
+        if(!notification.get().isDelivered()) {
+            this.schedulerService.updateScheduledNotification(newNotification);
+        }
+        return ConverterFactory.getNotificationConverter().entityToDto(newNotification);
+    }
+
+    @Transactional
+    public void removeNotificationsForUser(String projectId, String subjectId) {
+        Optional<Project> project = projectRepository.findByProjectId(projectId);
+
+        if(project.isEmpty()) {
+            throw new NotFoundException("Project not found with projectId " + projectId);
+        }
+
+        Optional<User> user = this.userRepository.findBySubjectIdAndProjectId(subjectId, project.get().getId());
+        if(user.isEmpty()) {
+            throw new NotFoundException("The supplied subject ID is invalid. No user found in the project ID provided. Please Create a User First.");
+        }
+
+        List<Notification> notifications = this.notificationRepository.findByUserId(user.get().getId());
+        this.schedulerService.deleteScheduledNotifications(notifications);
+
+        this.notificationRepository.deleteByUserId(user.get().getId());
+    }
+
+    @Transactional
+    public void updateDeliveryStatus(String fcmMessageId, boolean isDelivered) {
+        Optional<Notification> notification = this.notificationRepository.findByFcmMessageId(fcmMessageId);
+
+        notification.ifPresentOrElse(n -> {
+            n.setDelivered(isDelivered);
+            Notification newNotification = this.notificationRepository.save(n);
+        }, () -> {
+            throw new InvalidNotificationDetailsException("Notification with the provided FCM message ID does not exist.");
+        });
+    }
+
+    @Transactional
+    public void deleteNotificationByFcmMessageId(String fcmMessageId) {
+        this.notificationRepository.deleteByFcmMessageId(fcmMessageId);
+    }
+
+    public void removeNotificationsForUserUsingFcmToken(String fcmToken) {
+        Optional<User> user = this.userRepository.findByFcmToken(fcmToken);
+
+        user.ifPresentOrElse(user1 -> {
+            this.notificationRepository.deleteByUserId(user1.getId());
+            User newUser = user1.setFcmToken("");
```

Comment:
Does this mean you are deleting all the notifications of a user who has particular `fcmToken`?

Reply:
Yes. This happens when a a device is unregistered from Firebase (Uninstall/ Data clear of the app)